sw/device/lib/dif/dif_padctrl.c

```diff
@@ -0,0 +1,207 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include <stdbool.h>
+#include <stddef.h>
+
+#include "sw/device/lib/dif/dif_padctrl.h"
+
+// Generated.
+// Note that if the the number of DIO or MIO pads is such that all attribute can
+// fit in one register the PADCTRL_DIO_PADS0_REG_OFFSET and
+// PADCTRL_MIO_PADS0_REG_OFFSET #defines from padctrl_regs.h will no longer
+// contain '0' and become PADCTRL_DIO_PADS_REG_OFFSET and
+// PADCTRL_MIO_PADS_REG_OFFSET respectively which will require altering their
+// names in the source below.
```

Comment:
This behaviour is a little annoying, I've opened an issue to discuss here: https://github.com/lowRISC/opentitan/issues/2254

Do we alter reggen or do I do a local fix such as:

```C++
#ifdef PADCTRL_DIO_PADS_REG_OFFSET
  #define PADCTRL_DIO_PADS_REG_OFFSET_BASE PADCTRL_DIO_PADS0_REG_OFFSET
#else
  #define PADCTRL_DIO_PADS0_REG_OFFSET_BASE PADCTRL_DIO_PADS_REG_OFFSET
#endif   
```

Bearing in mind we may need such 'local' fixes in multiple places.

Reply:
We had a short discussion of this kind of DIF parameterisation before, when dealing with the PLIC, and I hoped the approach there would work with padctrl:

At the moment, we only care about the values of the parameters as will be instantiated in earlgrey, and not other possible instantiations. 

Given the software is making assumptions about these parameters, and we want the build to fail if it is not, the simplest way to ensure we reevaluate any assumptions is to have a series of `_Static_assert(…)`s at the top of `dif_padctrl.c`, using the `PADCTRL_PARAM_*` macros from `padctrl_regs.h`, like so:
```c
_Static_assert(PADCTRL_PARAM_NDIOPADS == 15, 
               "Padctrl instantiation parameters have changed.");
_Static_assert(PADCTRL_PARAM_NMIOPADS == 32, 
               "Padctrl instantiation parameters have changed.");
_Static_assert(PADCTRL_PARAM_ATTRDW == 8, 
               "Padctrl instantiation parameters have changed.");
```

This then means you can assume that these parameters have these values in the rest of the file, and that allows you to hard-code the correct version of `PADCTRL_DIO_PADS0_REG_OFFSET` or `PADCTRL_DIO_PADS_REG_OFFSET` in the short term.

In the longer term, yes it probably does make sense to change reggen to support this usecase, but that seems quite invasive, and may lead to lots of `0`s in register offset definitions which don't mean anything.

Reply:
Sounds good, do you think they should go in the tests .cc as well as the DIF .c? The current set of tests explicitly use all of the reggen #defines in a way that will break if the numbers change (e.g. see dio_attr_fields_ in dif_padctrl_test.cc) though as that will always be built with the DIF .c the static assert there may be sufficient?